src/compiler/checker.ts

```diff
@@ -38978,7 +38980,8 @@ namespace ts {
                         if (flags & ModifierFlags.Abstract) {
                             return grammarErrorOnNode(modifier, Diagnostics._0_modifier_already_seen, "abstract");
                         }
-                        if (node.kind !== SyntaxKind.ClassDeclaration) {
+                        // An abstract modifier is permitted on a class expression in a 'typeof abstract class {}' type
+                        if (node.kind !== SyntaxKind.ClassDeclaration && node.kind !== SyntaxKind.ClassExpression) {

```

Comment:
Wouldn't that also make it valid for a regular class expression? If not, could we consider allowing it in parse? It's never made sense to me that you can write this:
```ts
function f() {
  abstract class C {}
  return C;
}
```
But not this:
```ts
function f() {
  return abstract class C {};
}
```

Reply:
We certainly could allow it, but it would in effect be an expression syntax extension and we generally try to avoid those.